Some small performance fixes for c10 dispatcher

li-roy opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#20472 Some small performance fixes for c10 dispatcher**

Differential Revision: [D15332284](https://our.internmc.facebook.com/intern/diff/D15332284)

- Add path to call operator directly without creating kernel cache
- Save a refcount when getting dispatch key

Commit:
Some small performance fixes for c10 dispatcher

smessmer commented on aten/src/ATen/core/dispatch/DispatchTable.h:
```diff
@@ -217,8 +217,7 @@ class DispatchTable final {
       } else {
-        // TODO Avoid bumping the refcounter
-        return first_tensor_arg.toTensor().type_id();
+        return first_tensor_arg.unsafeToTensorImplRef()->type_id();
```
thanks :)

smessmer commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -133,6 +133,11 @@ class CAFFE2_API Dispatcher final {
+  /**
+   * Dispatch to and call the correct kernel directly, without a kernel cache
+   */
+  void call(const OperatorHandle& op, Stack* stack) const;
```
As discussed offline, let's remove this one because it isn't safe.

Commit:
Update on "Some small performance fixes for c10 dispatcher"

Some small performance fixes for c10 dispatcher

gh-metadata: pytorch pytorch 20472 gh/li-roy/18/head

Commit:
Update on "Some small performance fixes for c10 dispatcher"

Some small performance fixes for c10 dispatcher

gh-metadata: pytorch pytorch 20472 gh/li-roy/18/head

Commit:
Update on "Some small performance fixes for c10 dispatcher"

Some small performance fixes for c10 dispatcher

gh-metadata: pytorch pytorch 20472 gh/li-roy/18/head
